Spigot-Server-Patches/0234-Anti-Xray.patch

```diff
@@ -0,0 +1,1453 @@
+ 
++import java.util.Arrays;
+ import java.util.List;
+ 
++import com.destroystokyo.paper.antixray.ChunkPacketBlockControllerAntiXray.ChunkEdgeMode;
++import com.destroystokyo.paper.antixray.ChunkPacketBlockControllerAntiXray.EngineMode;
+ import net.minecraft.server.MinecraftServer;
+ import org.bukkit.Bukkit;
+ import org.bukkit.configuration.file.YamlConfiguration;
+@@ -428,4 +431,25 @@ public class PaperWorldConfig {
+         disableCreeperLingeringEffect = getBoolean("disable-creeper-lingering-effect", false);
+         log("Creeper lingering effect: " + disableCreeperLingeringEffect);
+     }
++
++    public boolean antiXray;
++    public boolean asynchronous;
++    public EngineMode engineMode;
++    public ChunkEdgeMode chunkEdgeMode;
++    public int maxChunkSectionIndex;
++    public List<Object> hiddenBlocks;
++    public List<Object> replacementBlocks;
++    private void antiXray() {
++        antiXray = getBoolean("anti-xray.enabled", false);
++        asynchronous = true;
++        engineMode = EngineMode.getById(getInt("anti-xray.engine-mode", 1));
++        engineMode = engineMode == null ? EngineMode.HIDE : engineMode;
++        chunkEdgeMode = ChunkEdgeMode.getById(getInt("anti-xray.chunk-edge-mode", 1));
++        chunkEdgeMode = chunkEdgeMode == null ? ChunkEdgeMode.DEFAULT : chunkEdgeMode;
++        maxChunkSectionIndex = getInt("anti-xray.max-chunk-section-index", 3);
++        maxChunkSectionIndex = maxChunkSectionIndex > 15 ? 15 : maxChunkSectionIndex;
++        hiddenBlocks = getList("anti-xray.hidden-blocks", Arrays.asList((Object) "gold_ore", "iron_ore", "coal_ore", "lapis_ore", "mossy_cobblestone", "obsidian", "chest", "diamond_ore", "redstone_ore", "lit_redstone_ore", "clay", "emerald_ore", "ender_chest"));
++        replacementBlocks = getList("anti-xray.replacement-blocks", Arrays.asList((Object) "stone", "planks"));
++        log("Anti-Xray: " + (antiXray ? "enabled" : "disabled") + " / Engine Mode: " + engineMode.getDescription() + " / Chunk Edge Mode: " + chunkEdgeMode.getDescription() + " / Up to " + ((maxChunkSectionIndex + 1) * 16) + " blocks");
++    }
+ }
+diff --git a/src/main/java/com/destroystokyo/paper/antixray/ChunkPacketBlockController.java b/src/main/java/com/destroystokyo/paper/antixray/ChunkPacketBlockController.java
+new file mode 100644
+index 00000000..7e6481c1
+--- /dev/null
++++ b/src/main/java/com/destroystokyo/paper/antixray/ChunkPacketBlockController.java
+@@ -0,0 +1,44 @@
++package com.destroystokyo.paper.antixray;
++
++import net.minecraft.server.BlockPosition;
++import net.minecraft.server.Chunk;
++import net.minecraft.server.IBlockData;
++import net.minecraft.server.PacketPlayOutMapChunk;
++import net.minecraft.server.World;
++
++public class ChunkPacketBlockController {
++
++    private static ChunkPacketBlockController instance = null;
++
++    protected ChunkPacketBlockController() {
++        
++    }
++
++    public static synchronized ChunkPacketBlockController getInstance() {
```

Comment:
There is no point for this being synchronized if this class holds no mutable data. Set the field to static final and be done with it.

Note that final fields are treated specially in the JLS so no synchronization is required for their use.

`private static final ChunkPacketBlockController instance = new ChunkPacketBlockController() ;`